### PR TITLE
feat(CenterPoint): support multi-task interface in the new CenterHead design

### PR DESCRIPTION
**Pull request by KSeangTan**
## Summary
This PR requires the changes in this [PR](https://github.com/tier4/autoware-ml/pull/86), where it takes the changes and implement them to the new `CenterHead`. Besides, it also adds few dataclasses to structure and clean up inputs/outputs from models. Therefore, downstream tasks/functions can infer their outputs easily. 

## Main Changes:
- Add `CenterHead` to support the multi-task interface, where the main changes are in `get_targets`[[1]](https://github.com/KSeangTan/autoware-ml/blob/7e74990eda00bb0059c220b51b7cc97ca06619bc/autoware_ml/models/detection3d/heads/centerhead.py#L155), and `decode_outputs`[[2]](https://github.com/KSeangTan/autoware-ml/blob/7e74990eda00bb0059c220b51b7cc97ca06619bc/autoware_ml/models/detection3d/heads/centerhead.py#L481) that their implementations are mostly in vectorization.
-  Add `MultiTaskPredictions` [[3]](https://github.com/tier4/autoware-ml/pull/87/changes#diff-7d8ba5395265ffc688fb8018cbb6286b6da16bd6d654407b7ceb5eba33ea6ffcR14) to save decoded predictions from different tasks.
- Add `TransFusionHeadOutputs` [[4]](https://github.com/tier4/autoware-ml/pull/87/changes#diff-09c6b0713c673b31cca6cc3c1b88e593a02fcf55ca5e24af0e3d9982f1d64f52R11) and `CenterHeadOutputs` [[5]](https://github.com/tier4/autoware-ml/pull/87/changes#diff-09c6b0713c673b31cca6cc3c1b88e593a02fcf55ca5e24af0e3d9982f1d64f52R29) to save raw outputs from different networks.
- Add `Detection3DPredictions` [[6]](https://github.com/tier4/autoware-ml/pull/87/changes#diff-f337797f9cbef110616017526d3d60207c549bce655f2168b9d8a1da02dc4f39R11) to save decoded predictions for detection3d.
- Add `CenterHeadTargets` [[7]](https://github.com/tier4/autoware-ml/pull/87/changes#diff-ecdb1da508dbdf60a787ec43eff17a3217f19a75ccacd906b2e76c87854ac6f1R11) to save encoded targets for `CenterHead`.
- Add unittests for `CenterHead` [[8]](https://github.com/tier4/autoware-ml/pull/87/changes#diff-a2ae1db09433bf67aadb91a0e7e1e26815f52020eeab92e1b8ed4e661bcb116fR1).

## Checklist

- [x] You've checked our [contribution overview](https://tier4.github.io/autoware-ml/contributing/contribution-overview/).
- [x] Your PR follows our [pull request guidelines](https://tier4.github.io/autoware-ml/contributing/pull-request-guidelines/).

## Unit Tests
```
python -m unittest autoware_ml/models/detection3d/heads/tests/test_centerhead.py 
............
----------------------------------------------------------------------
Ran 12 tests in 0.538s

OK
```

### CODE REVIEW

File: autoware_ml/models/detection3d/heads/centerhead.py [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+        )
+
+        if self.use_velocity:
+            vel_targets = gt_bboxes_3d[
+                :, :, Box3DFieldIndex.VELOCITY_X : Box3DFieldIndex.VELOCITY_Y + 1
+            ]
+            reg_targets = torch.cat([reg_targets, vel_targets], dim=-1)
+
+        # (batch_size, max_num_bboxes) -> (batch_size, max_num_bboxes)
+        reg_indices = (center_int[:, :, 1] * feature_width + center_int[:, :, 0]).long()
```
Comment (vividf) [inline autoware_ml/models/detection3d/heads/centerhead.py:267]: Out-of-range GT centers still produce reg_indices, which are passed to gather() before the validity mask is applied. This can create out-of-bounds indices and crash training. Please sanitize/clamp the gather indices first, while keeping valid_bbox_masks to exclude those entries from the loss.
```
out-of-range GT
      ↓
valid_bbox_masks = False
      ↓
calculate illegal reg_indices
      ↓
gather()  ←crash here
      ↓
multiply valid mask
```

File: autoware_ml/models/detection3d/heads/centerhead.py [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+        device: torch.device,
+    ) -> CenterHeadTargets:
+        """Build heatmap and regression targets for one batch."""
+        batch_size = len(gt_bboxes_3d)
+        max_num_bboxes = gt_bboxes_3d.shape[1]
+        feature_height, feature_width = feature_map_size
+
+        # Movement of tensors to the correct device and type
+        # Get only the first K params for ground truths
+        gt_bboxes_3d[:, :, : self.box_code_size] = gt_bboxes_3d[:, :, : self.box_code_size].to(
```
Comment (vividf) [inline autoware_ml/models/detection3d/heads/centerhead.py:174]: This does not actually move gt_bboxes_3d itself to the target device and also mutates the input tensor in place. Please use gt_bboxes_3d = gt_bboxes_3d.to(device) instead.

File: autoware_ml/models/detection3d/heads/centerhead.py [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+        )
+        loss_heatmap = self.loss_heatmap(output_heatmaps, targets.heatmaps)
+
+        bbox_predictions = [
+            outputs.center_head_outputs.centers,
+            outputs.center_head_outputs.heights,
+            outputs.center_head_outputs.dims,
+            outputs.center_head_outputs.rots,
+        ]
+        if self.use_velocity and outputs.center_head_outputs.vels is not None:
```
Comment (vividf) [inline autoware_ml/models/detection3d/heads/centerhead.py:321]: When use_velocity=True, vels=None leaves an 8-channel prediction tensor but get_targets() still builds 10-channel regression targets, which will fail later with a shape mismatch. Please validate the output contract explicitly.

File: autoware_ml/models/detection3d/heads/centerhead.py [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+        )
+
+        # (batch_size, max_num_bboxes) boolean mask for valid boxes based on the number of valid boxes per sample
+        # (max_num_bboxes) -> (1, max_num_bboxes) -> (batch_size, max_num_bboxes) < gt_valid_bboxes.unsqueeze(1) (batch_size, 1)
+        # -> (batch_size, max_num_bboxes)
+        valid_num_bboxes_masks = torch.arange(max_num_bboxes, device=device).unsqueeze(0).expand(
+            batch_size, -1
+        ) < gt_valid_bboxes.unsqueeze(1)
+        # (batch_size, max_num_bboxes) boolean mask for valid boxes based on both distance and number of valid boxes
+        valid_bbox_masks = valid_distance_masks & valid_num_bboxes_masks
```
Comment (vividf) [inline autoware_ml/models/detection3d/heads/centerhead.py:207]:
```suggestion
valid_label_masks = (
    (gt_labels_3d >= 0)
    & (gt_labels_3d < self.num_classes)
)

valid_bbox_masks = (
    valid_distance_masks
    & valid_num_bboxes_masks
    & valid_label_masks
)
```

**Review by vividf — APPROVED** [review]
LGTM! Left some comments

File: autoware_ml/models/detection3d/heads/centerhead.py (hunk 4 above)
Comment (KSeangTan) [reply autoware_ml/models/detection3d/heads/centerhead.py:207]: Thanks, we don't actually need to care valid_label_masks here, all invalid label bboxes should be explicitly validated beforehand in the transform pipeline, and update `gt_valid_bboxes`, e.g., `BBoxesLabelNameFilter`.

**Review by amadeuszsz — APPROVED** [review]
@KSeangTan  Thank you for this PR.

This PR moves forwards a new database format. However, it introduces code duplication. We can merge this PR, but in the following PR we will need to move this implementation to new location and delete the old code. FYI I'm already working on base class refactor and PTv3 refactor so we can remove duplicated code soon (PTv3 is blocker because for now it is the only one model actively developed in Autoware-ML).

**Comment by KSeangTan** [comment]
Thanks @amadeuszsz 

I am actually working in another PR to move all models to the new interface, so i plan to create another list of stack PRs to review and merge that. For now, we can actually close this PR and not merging it.

**Comment by KSeangTan** [comment]
Close the PR
